### PR TITLE
Fix caching of union constituent instantiations that share a declaration

### LINKED ISSUE #40995: typescript generic break filter after spreading object in map
**TypeScript Version:**  any, including 4.0.2 and v4.1.0-dev.20201008

**Search Terms:** typescript generic break filter after spreading object in map? O_o

**Code**

```ts
interface Success {
    isSuccess: true;
}

interface Fail {
    isSuccess: false;
}

type Item =
    | Success
    | Fail

// ok
function foo(a: Array<Item>): void {
    a
        .map((item) => ({...item}))
        .filter((value) => {
            console.log(value);
        });
}

// error
function bar<T extends any>(a: Array<Item>): void {
    a
        .map((item) => ({...item}))
        .filter((value) => {
            console.log(value);
        });
}
```

**Expected behavior:**

adding generic to function does not break the code

**Actual behavior:**

`.filter((value) => {` after `.map((item) => ({...item}))` when it inside function with generic trigger error.

**Playground Link:** https://www.typescriptlang.org/play?ts=4.0.2#code/JYOwLgpgTgZghgYwgAgMoFcFIM7eQbwChkTlhsMsJcAuZMKdCAbkIF9DDRJZEUAxOMAA2BYqXKUc2OvGHYW7TmACeABxQBJSAFtkAXnEkAPmkzSjyU4JGcA9HeQB7ANaEY6EAjDAnIZDBOTgAUcHQAglBQcCoAPNoQOgB8AJR0AG5OwAAmYqTIcJb5AHQ6cGrBwcC6KQZJyMH4xc3ViWwpKUWkxTAiPJXpcMJMtfr1RPmTpAh+2E7CEMXCTgDmwYPDECnMyA7IG0x0CTpdJO2sHIR70FBOUO6e3r7+AEZwULEAKsgQAB6QIGyeDgIBUSVCESiMXiulSGSyuQmpEKUxIpXKlVaOlG9UazWKWPanVRyB6fWgAyGIzqeRJ01m80WyzWBy2Oz2AH4OadkOclEA

### PR DESCRIPTION
**Pull request by andrewbranch**
Not ready for review.

Fixes #40995

### CODE REVIEW

**Comment by andrewbranch** [comment]
@typescript-bot perf test this

**Comment by andrewbranch** [comment]
Of course, a _huge_ increase for material-ui, <0.5% difference everywhere else 🤦

**Comment by andrewbranch** [comment]
@typescript-bot perf test this
@typescript-bot pack this

**Comment by andrewbranch** [comment]
This is now strictly better than the current behavior, and has identical caching/performance characteristics, but I’m unsure whether it’s a complete fix. (That is, there may be more cases where returning a cached instantiation is wrong.)

**Comment by ahejlsberg** [comment]
@andrewbranch It's not immediately obvious to me what's going on here. Would you mind explaining what the core issue is?

**Comment by andrewbranch** [comment]
I’ll describe what happens _before_ this change. This code errors because “No overload matches this call” on the `.filter(...)`:

```ts
interface Success {
    isSuccess: true;
}

interface Fail {
    isSuccess: false;
}

function bar<Q>(a: (Success | Fail)[]) {
    a
        .map(item => ({...item}))
        .filter(value => {
            return value.isSuccess;
        });
}
```

The error elaboration for the overload that ought to match is

```
Overload 2 of 2, '(predicate: (value: { isSuccess: true; } | { isSuccess: false; }, index: number, array: ({ isSuccess: true; } | { isSuccess: false; })[]) => unknown, thisArg?: any): ({ isSuccess: true; } | { isSuccess: false; })[]', gave the following error.
    Argument of type '(value: { isSuccess: true; }) => true' is not assignable to parameter of type '(value: { isSuccess: true; } | { isSuccess: false; }, index: number, array: ({ isSuccess: true; } | { isSuccess: false; })[]) => unknown'.
      Types of parameters 'value' and 'value' are incompatible.
        Type '{ isSuccess: true; } | { isSuccess: false; }' is not assignable to type '{ isSuccess: true; }'.
          Type '{ isSuccess: false; }' is not assignable to type '{ isSuccess: true; }'.
            Types of property 'isSuccess' are incompatible.
              Type 'false' is not assignable to type 'true'.
```

and quick info on `value` shows `{ isSuccess: true }` instead of the union, providing a clue as to what’s going wrong.

The first important thing that happens is the spread in the `.map(...)` call creates new type identities that are structurally identical to `Success` and `Fail` and `Success | Fail` (which is why printback no longer uses those names afterward), and both the new `{ isSuccess: true }` and the new `{ isSuccess: false }` types share their declaration site at `{ ...item }`.

Where the problem arises is during signature resolution for `.filter(...)`. We try on the first overload, which is

```ts
interface Array<T> {
  // ...
  filter<S extends T>(predicate: (value: T, index: number, array: T[]) => value is S, thisArg?: any): S[];
  // ...
}
```

and try to infer `S`. Since our predicate is not a type predicate, inference initially produces `unknown`, which we will need to replace with the constraint of `T`, which is `{ isSuccess: true } | { isSuccess: false }`. We first instantiate this constraint with the inference context’s non-fixing mapper, which ought to act as the identity function since this type does not reference `S`, but it goes off the rails.

We instantiate each union constituent through `getObjectTypeInstantiation` (we’ve finally made it to the code this PR changes). Recall that _each_ constituent’s symbol’s declaration is `{ ...item }`. We have an outer type parameter `Q`, which should have no effect on the instantiation in the end, but we don’t determine that without trying it. But first, we cache our union constituent `{ isSuccess: true }` as the result of the instantiation for `Q` on the node links for `{ ...item }` (this is the line I removed). When we move on to the second union constituent, `{ isSuccess: false }`, we have a problem: since the declaration is the same, we already have `links.outerTypeParameters` cached, so we move on to the real work of the function (`if (typeParameters.length) { ... }`). We get the `typeArguments` for this instantiation by running the outer type parameters (`Q`) through the inference context’s mapper. Because `Q` was not the inference target, it returns `Q`, so our type arguments are just `Q`. We look up the instantiation for `Q` on the links, and get the result from the previous union constituent, `{ isSuccess: true }`—clearly wrong.

So in short, the union `{ isSuccess: true } | { isSuccess: false }` collapses to the first constituent during an instantiation that should have been a no-op because the first one gets cached as the instantiation of an unused outer type parameter on their shared declaration.

**Comment by ahejlsberg** [comment]
Hmm, that's pretty subtle. The issue is that the logic in `getObjectTypeInstantiation` implicitly assumes only a single type is associated with the declaration. That was always the case when the logic was written, but with union type distribution over spreads in object literals it isn't anymore. The fix is to move the instantiation cache from the `NodeLinks` associated with the declaration to the `target` object type being instantiated. That'll allow multiple types to share the same declaration, but we'll still have a single shared cache for each type. I tried it out and it seems to work great. I'll put up a PR that you can poach from.

**Comment by ahejlsberg** [comment]
#41084 is what I think is the correct fix.

**Comment by andrewbranch** [comment]
> The fix is to move the instantiation cache from the NodeLinks associated with the declaration to the `target` object type being instantiated.

Ah, the first thing I tried was effectively equivalent to moving the cache to the `type`, which in my test case was the same as `target`, but defeated the purpose of the cache when an actual generic type had many instantiations (giving us the bad perf on material-ui). This makes more sense. Thanks!
